### ISSUE TITLE
range_minus_one changes range type and causes type error

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9908]
### Summary

.

### Lint Name

range_minus_one

### Reproducer

I tried this code:

```rust
fn main() {
    // Simplified test case
    let _ = || 0..=1;

    // Original test case
    let full_length = 1024;
    let range = {
        // do some stuff, omit here
        None
    };

    let range = range.map(|(s, t)| s..=t).unwrap_or(0..=(full_length-1));

    assert_eq!(range, 0..=1023);
}
```

I saw this happen:
`cargo clippy --fix -- -Aclippy::all -Wclippy::range_minus_one`
```
after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
   --> src/main.rs:13:53
    |
13  |     let range = range.map(|(s, t)| s..=t).unwrap_or(0..full_length);
    |                                           --------- ^^^^^^^^^^^^^^ expected struct `std::ops::RangeInclusive`, found struct `std::ops::Range`
    |                                           |
    |                                           arguments to this function are incorrect
    |
    = note: expected struct `std::ops::RangeInclusive<_>`
               found struct `std::ops::Range<{integer}>`
note: associated function defined here
   --> /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:799:18
    |
799 |     pub const fn unwrap_or(self, default: T) -> T
    |                  ^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: an exclusive range would be more readable
  --> src/main.rs:13:53
   |
13 |     let range = range.map(|(s, t)| s..=t).unwrap_or(0..=(full_length-1));
   |                                                     ^^^^^^^^^^^^^^^^^^^ help: use: `0..full_length`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#range_minus_one
   = note: requested on the command line with `-W clippy::range-minus-one`

warning: `clpy` (bin "clpy" test) generated 1 warning (run `cargo fix --bin "clpy" --tests` to apply 1 suggestion)
warning: `clpy` (bin "clpy") generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.67s
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by KnorpelSenf** [comment]
I am running into the same issue with the following code.

```rs
#![deny(clippy::range_minus_one)]

fn main() {
    foo(12);
}

fn foo(n: usize) -> usize {
    (1..=n - 1).sum()
}
```

It leads to the following output.

```rs
error: an exclusive range would be more readable
 --> src/main.rs:8:5
  |
8 |     (1..=n - 1).sum()
  |     ^^^^^^^^^^^ help: use: `1..n`
```

If `--fix` is supplied, the following error is shown.

```rs
The following errors were reported:
error[E0689]: can't call method `sum` on type `usize`
 --> src/main.rs:8:10
  |
8 |     1..n.sum()
  |          ^^^ can't call method `sum` on type `usize`
  |
help: you must surround the range in parentheses to call its `sum` function
  |
8 |     (1..n).sum()
  |     +    +

error[E0308]: mismatched types
 --> src/main.rs:8:5
  |
7 | fn foo(n: usize) -> usize {
  |                     ----- expected `usize` because of return type
8 |     1..n.sum()
  |     ^^^^^^^^^^ expected `usize`, found `Range<{integer}>`
  |
  = note: expected type `usize`
           found struct `std::ops::Range<{integer}>`

error: aborting due to 2 previous errors
```

**Comment by samueltardieu** [comment]
> I am running into the same issue with the following code.

This is not the same issue, but both will be fixed by #14432.

**Comment by KnorpelSenf** [comment]
@samueltardieu that's actually true, I didn't read closely enough. Would you like me to move my comment to a new issue? #14432 could then be edited to also close that new issue.

**Comment by samueltardieu** [comment]
No need, this is fine. The commit messages detail what is fixed, and contain non-regression tests, so the issue number is not important.
